File: src/Sample/Plumix.Sample/Demos/General/TrackpadPanZoomDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/trackpad_pan_zoom_demo_page.dart (exact sample parity)

public sealed class TrackpadPanZoomDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new TrackpadPanZoomDemoPageState();
    }
}

internal sealed class TrackpadPanZoomDemoPageState : State
{
    private Point _pan;
    private double _scale = 1.0;
    private double _rotation;
    private int _updates;
    private string _phase = "idle";

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("Trackpad pan / zoom", fontSize: 20, color: Colors.Black),
                new Text(
                    "Pinch or rotate on a trackpad over the panel. The platform reports the gesture "
                    + "as PointerPanZoom events, which Listener surfaces directly.",
                    fontSize: 14,
                    color: Colors.DimGray),
                BuildProbe(),
                new Text(
                    $"phase {_phase} — pan {Format(_pan.X)}, {Format(_pan.Y)} — "
                    + $"scale {Format(_scale)} — rotation {Format(_rotation)} rad — {_updates} updates",
                    fontSize: 13,
                    color: Colors.DimGray),
            ]);
    }

    private Widget BuildProbe()
    {
        return new Listener(
            behavior: HitTestBehavior.Opaque,
            onPointerPanZoomStart: _ => SetState(() =>
            {
                _phase = "active";
                _pan = default;
                _scale = 1.0;
                _rotation = 0.0;
                _updates = 0;
            }),
            onPointerPanZoomUpdate: @event => SetState(() =>
            {
                _pan = @event.Pan;
                _scale = @event.Scale;
                _rotation = @event.Rotation;
                _updates++;
            }),
            onPointerPanZoomEnd: _ => SetState(() => _phase = "idle"),
            child: new Container(
                height: 220,
                alignment: Alignment.Center,
                decoration: new BoxDecoration(
                    Color: Color.Parse("#FFF1F3F4"),
                    Border: Rendering.Border.FromBorderSide(new BorderSide(
                        color: Color.Parse("#FF9AA0A6"),
                        width: 1)),
                    BorderRadius: BorderRadius.Circular(10)),
                child: BuildTarget()));
    }

    private Widget BuildTarget()
    {
        Matrix4 transform = Matrix4.TranslationValues(_pan.X, _pan.Y, 0.0);
        transform = transform.Multiplied(Matrix4.RotationZ(_rotation));
        transform = transform.Multiplied(Matrix4.Diagonal3Values(_scale, _scale, 1.0));

        return new Widgets.Transform(
            transform: transform,
            alignment: Alignment.Center,
            child: new Container(
                width: 96,
                height: 96,
                alignment: Alignment.Center,
                decoration: new BoxDecoration(
                    Color: Color.Parse("#FF00796B"),
                    BorderRadius: BorderRadius.Circular(12)),
                child: new Text("pinch me", fontSize: 13, color: Colors.White)));
    }

    private static string Format(double value)
    {
        return value.ToString("0.00", System.Globalization.CultureInfo.InvariantCulture);
    }
}

File: dart_sample/lib/demos/general/trackpad_pan_zoom_demo_page.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/widgets.dart';

class TrackpadPanZoomDemoPage extends StatefulWidget {
  const TrackpadPanZoomDemoPage({super.key});

  @override
  State<TrackpadPanZoomDemoPage> createState() => _TrackpadPanZoomDemoPageState();
}

class _TrackpadPanZoomDemoPageState extends State<TrackpadPanZoomDemoPage> {
  Offset _pan = Offset.zero;
  double _scale = 1.0;
  double _rotation = 0.0;
  int _updates = 0;
  String _phase = 'idle';

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Trackpad pan / zoom',
          style: TextStyle(fontSize: 20, color: Color(0xFF000000)),
        ),
        const Text(
          'Pinch or rotate on a trackpad over the panel. The platform reports the gesture '
          'as PointerPanZoom events, which Listener surfaces directly.',
          style: TextStyle(fontSize: 14, color: Color(0xFF696969)),
        ),
        _buildProbe(),
        Text(
          'phase $_phase — pan ${_format(_pan.dx)}, ${_format(_pan.dy)} — '
          'scale ${_format(_scale)} — rotation ${_format(_rotation)} rad — $_updates updates',
          style: const TextStyle(fontSize: 13, color: Color(0xFF696969)),
        ),
      ],
    );
  }

  Widget _buildProbe() {
    return Listener(
      behavior: HitTestBehavior.opaque,
      onPointerPanZoomStart: (PointerPanZoomStartEvent event) => setState(() {
        _phase = 'active';
        _pan = Offset.zero;
        _scale = 1.0;
        _rotation = 0.0;
        _updates = 0;
      }),
      onPointerPanZoomUpdate: (PointerPanZoomUpdateEvent event) => setState(() {
        _pan = event.pan;
        _scale = event.scale;
        _rotation = event.rotation;
        _updates++;
      }),
      onPointerPanZoomEnd: (PointerPanZoomEndEvent event) => setState(() => _phase = 'idle'),
      child: Container(
        height: 220,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: const Color(0xFFF1F3F4),
          border: Border.all(color: const Color(0xFF9AA0A6)),
          borderRadius: BorderRadius.circular(10),
        ),
        child: _buildTarget(),
      ),
    );
  }

  Widget _buildTarget() {
    final Matrix4 transform = Matrix4.translationValues(_pan.dx, _pan.dy, 0.0)
      ..multiply(Matrix4.rotationZ(_rotation))
      ..multiply(Matrix4.diagonal3Values(_scale, _scale, 1.0));

    return Transform(
      transform: transform,
      alignment: Alignment.center,
      child: Container(
        width: 96,
        height: 96,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: const Color(0xFF00796B),
          borderRadius: BorderRadius.circular(12),
        ),
        child: const Text(
          'pinch me',
          style: TextStyle(fontSize: 13, color: Color(0xFFFFFFFF)),
        ),
      ),
    );
  }

  static String _format(double value) => value.toStringAsFixed(2);
}
